### PR TITLE
Allow display name changes in data explorer (1702)

### LINKED ISSUE #1701: Issue with slashes in data titles in CanSAS1D (and probably NXcanSAS)
Up to and including release 5.0.3, the Data Explorer displayed the name of the _file_ that had been loaded. Since the release, several improvements and bug fixes have been made to the data loader. One improvement being that if it can recognise the name of the _data_ in the file, it will display this in the Data Explorer instead of the file name. The CanSAS1D and NXcanSAS formats will benefit most from this.

In testing the fix to https://github.com/SasView/sasview/pull/1683, User Adrian R has spotted that if the string between <title> </title> contains a forward slash or a back slash, none of the string before the slash gets displayed as the data name in the Data Explorer. The full data name is, however, displayed when you expand Info for that data set. Thus it would seem that the code translating the title string into a data set name is interpreting the slashes.

Characters like percent or vertical bar are unaffected.

The following zip file contains a screenshot illustrating the issue and some example datasets.
[TitleAnomaly.zip](https://github.com/SasView/sasview/files/5465790/TitleAnomaly.zip)

**NB: You must run a recent developer build to see this behaviour! For example:**

https://jenkins.esss.dk/sasview-beta/job/SasView-Windows10-ghprb/58/artifact/sasview/dist/setupSasView.exe after Oct 27.

### LINKED ISSUE #1702: Allow for a choice of how data is named in the Data Explorer
See also https://github.com/SasView/sasview/issues/1701.

Up to and including release 5.0.3, the Data Explorer displayed the name of the file that had been loaded. Since the release, several improvements and bug fixes have been made to the data loader. One improvement being that if it can recognise the name of the data in the file, it will display this in the Data Explorer instead of the file name. The CanSAS1D and NXcanSAS formats will benefit most from this.

User Adrian R makes the following comment on this switch:
```
Perhaps I can make a suggestion about this enhancement in version 5: while I realise that often people like the sample 
description as provided in the title of a data file to appear in plots or in the table when selecting data, there are 
equally cases where one does not find that appropriate.   A choice would be useful.

For example, there can be repeated runs for the same sample, either as a time sequence with varying properties/sample 
conditions or simply made as checks that data are reproduced.  In this case the title may be identical for several data 
sets but the measurements are distinguished by file name or run/sequence number.  It could then be convenient to 
choose what is presented both in the data panel and on the plot/fit output.

The data files that I sent [see #1701 ] happened to be some where the displaying by title could be appropriate but with 
part of the information missing, this became much less helpful.  There are quite a number of data sets where I have 
several measurements with the same title that would only be distinguished by other metadata (instrument configuration, 
wavelength, ....).
```

### PR DESCRIPTION
**Pull request by krzywon**
A new option was added to the data explorer context menu allowing users to change the display name for a data set. The option opens a window offering 4 options: Keep the existing name, change to the data tile, change to the data file name, or change to any name the user would like. Unit tests were added to check base functionality and catch a few odd edge cases.

This fixes #1701 and closes #1702.

### CODE REVIEW

File: src/sas/qtgui/MainWindow/NameChanger.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        """
+        Find the radio button that is selected and find its associated textbox
+        """
+        buttonStates = [self.rbExisting.isChecked(), self.rbDataName.isChecked(),
+                   self.rbFileName.isChecked(), self.rbNew.isChecked()]
+        textValues = [self.txtCurrentName.text(), self.txtDataName.text(),
+                      self.txtFileName.text(), self.txtNewCategory.text()]
+        newValues = [textValues[i] for i, value in enumerate(textValues) if buttonStates[i]]
+        name = newValues[0] if len(newValues) == 1 and newValues[0] else self.txtCurrentName.text()
+        self._model_item.setText(name)
```
Comment (rozyczko) [inline src/sas/qtgui/MainWindow/NameChanger.py:72]: There is no check on name collision - do we want to allow multiple items with the same name?
Currently, when a dataset is loaded with the same name, it gets the standard extension [1], [2] etc.

**Review by rozyczko — CHANGES_REQUESTED** [review]
Code looks good, minor functionality related issues raised.

File: src/sas/qtgui/MainWindow/DataExplorer.py [hunk 2]
```diff
@@ -1498,6 +1505,21 @@ def onCustomContextMenu(self, position):
         # Fire up the menu
         self.context_menu.exec_(self.current_view.mapToGlobal(position))
 
+    def changeName(self):
+        """
+        Open a modal window that can change the display name of the selected data
```
[the remark this answers is not in the document]
Comment (krzywon) [reply src/sas/qtgui/MainWindow/DataExplorer.py:1510]: This was supposed to be a modal window. I should add a data deletion signal to close the window when the model item is deleted. The second case isn't likely if the window is modal, but I would prefer to cover edge cases.

File: src/sas/qtgui/MainWindow/NameChanger.py (hunk 1 above)
Comment (krzywon) [reply src/sas/qtgui/MainWindow/NameChanger.py:72]: Good catch - I'll pass the new name through the DataManager rename routine

**Comment by krzywon** [comment]
5.0 ready for testing on Win

File: src/sas/qtgui/MainWindow/NameChanger.py [hunk 3]
```diff
@@ -71,6 +81,16 @@ def getNewText(self):
...
+    def removeData(self, data_list=None):
+        """
+        Safely remove data from the window in the unlikely event a data deletion signal is sent to the modal window
+        """
+        if not data_list or self._model_item not in data_list:
+            return
+        # Reset model_item and data to None and close the window
+        self.model_item = None
+        self.close()
+
```
Comment (rozyczko) [inline src/sas/qtgui/MainWindow/NameChanger.py:100]: Very nice feature! We should have these safeguards in more places.

**Review by rozyczko — APPROVED** [review]

**Comment by rozyczko** [comment]
Changes look good, all issues addressed. Ready to merge

**Comment by krzywon** [comment]
Here are a few comments I've received from Adrian Rennie on the functionality. I'm going to try to implement his suggestions but I'm not sure I will be able to in the PR.

> (a)  One of the main reasons to use this option is when there are several files that appear, with for example, the same title.  In this case, it would be useful to select a number of runs and choose an option for update automatically to e.g. the filename if that is distinctive and the etitle is always the same.
>
>(b)  A further comment is that if one changes the name and then changes it back again to what it was previously a '(1)' appears after the name although there is not any extra data set showing in the Data Explorer.  Is this intentional? 
>
>(c)  In terms of future use, I could imagine that people might like to use some automatic selection as the name of other information that might be available.  At the moment that is usually rather little but might include the 'Run' number but one might envisage loading other metadata as parameters from a canSAS or nxs file.

**Comment by rozyczko** [comment]
> > (b)  A further comment is that if one changes the name and then changes it back again to what it was previously a '(1)' appears after the name although there is not any extra data set showing in the Data Explorer.  Is this intentional?

This looks like a bug, the state should know about data filenames at current time. 

> > (c)  In terms of future use, I could imagine that people might like to use some automatic selection as the name of other information that might be available.  At the moment that is usually rather little but might include the 'Run' number but one might envisage loading other metadata as parameters from a canSAS or nxs file.

This is something we have thought about before and even created a ticket (here or locally in Jira, don't recall).
Modifying the visibility of datasets (not just the check status) based on various criteria (datatype, string in name etc)
Definitely (c) is a separate, large project.

**Comment by smk78** [comment]
Have just been testing this branch.
I note that when you plot data, the plot legend always displays the 'data name' irrespective of the 'filename' or any name assigned. However, fitting models use the name assigned, not the data name. For example:
image
image
Here I left the name of the core contrast unmodified, but called the drop contrast 'fred'.
I think this is likely to be very confusing?

**Comment by butlerpd** [comment]
The point raised by @smk78 has been addressed in another ticket #1616 and a PR #1745 while c) needs to be turned into another ticket. a) looks like an extra feature request. Is that a quick fix or should it be a separate ticket?

That leaves b) as a bug to address?
